pkg/virt-controller/watch/vm.go

```diff
@@ -271,7 +271,110 @@ func (c *VMController) execute(key string) error {
 		return err
 	}
 
-	return createErr
+	if createErr != nil {
+		return createErr
+	}
+
+	err = handleVMRenaming(vm, c, logger)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleVMRenaming(vm *virtv1.VirtualMachine, c *VMController, logger *log.FilteredLogger) error {
+	// Check if renaming annotation exists
+	if vm.ObjectMeta.Annotations == nil {
+		return nil
+	}
+
+	renameTo, hasRenameToAnnotation := vm.ObjectMeta.Annotations[virtv1.RenameToAnnotation]
+
+	// Handle source VM
+	if hasRenameToAnnotation {
+		// If a VM with the new name exists, remove annotation
+		_, err := c.clientset.VirtualMachine(vm.Namespace).Get(renameTo, &v1.GetOptions{})
+
+		if err != nil {
+			if !strings.Contains(err.Error(), "not found") {
+				errMsg :=
+					fmt.Sprintf("can't determine existence of newVM: %s while processing rename of oldVM: %s",
+						renameTo, vm.Name)
+				logger.Reason(err).Error(errMsg)
+				return err
+			} else {
+				// New vm is not found, create it
+				newVM := &virtv1.VirtualMachine{
+					TypeMeta: vm.TypeMeta,
+					ObjectMeta: v1.ObjectMeta{
+						Name:      renameTo,
+						Namespace: vm.Namespace,
+						Labels:    vm.Labels,
+						Annotations: map[string]string{
+							virtv1.RenameFromAnnotation: vm.Name,
+						},
+						// Finalizers: vm.Finalizers,
+					},
+					Spec: vm.Spec,
+				}
+
+				_, err := c.clientset.VirtualMachine(vm.Namespace).Create(newVM)
+
+				if err != nil {
+					errMsg := fmt.Sprintf("Could not create a new vm during the rename process of: %s", vm.Name)
+					logger.Reason(err).Error(errMsg)
+					return err
+				}
+			}
+		}
+
+		// New VM exists, remove renameTo annotation
+		patchString := fmt.Sprintf("{\"metadata\":{\"annotations\":{\"%s\":null}}}",
+			virtv1.RenameToAnnotation)
+		_, err = c.clientset.VirtualMachine(vm.Namespace).Patch(vm.Name, types.MergePatchType, []byte(patchString))
+
+		if err != nil {
+			logger.Reason(err).Error(fmt.Sprintf("Failed removing 'renameTo' annotation from vm: %s", vm.Name))
+			return err
+		}
+
+		return nil
+	}
+
+	renameFrom, hasRenameFromAnnotation := vm.ObjectMeta.Annotations[virtv1.RenameFromAnnotation]
+
+	// Handle destination VM
+	if hasRenameFromAnnotation {
+		// Attempt to delete the old vm
+		err := c.clientset.VirtualMachine(vm.Namespace).Delete(renameFrom, &v1.DeleteOptions{})
+
+		if err != nil {
+			if !strings.Contains(err.Error(), "not found") {
+				errMsg :=
+					fmt.Sprintf("can't delete oldVM: %s while processing newVM: %s",
```

Comment:
I think that we should hide implementation details from the user... as far as the user knows, he asked us to rename a VM, not create new and remove the old one...
The error should at least tell the user  that the rename process failed...

Reply:
You are right about hiding implementation details, but the rename process did not fail at this point, only this attempt at the process has failed
The VM would be re-enqueued, and the controller would try again. Notice that this code is reached when an ambiguous error occurs (`err != 'not found'`), meaning I cannot be certain that the old VM was actually deleted, so I have to re-enqueque the VM untill I successfully delete it or receive `err == 'not found'`, only then I am certain it was actually deleted.
What's an error message you would recommend? I don't want to write vague error messages like "Rename process failed, will retry" because it doesn't hint to what the exact problem was in the process

Reply:
I've modified the error messages